File: DSA/DSA08003 HÀNG ĐỢI HAI ĐẦU (DEQUEUE).cpp
#include <bits/stdc++.h>
using namespace std;

void submit()
{
    int n;
    cin >> n;
    cin.ignore();
    deque<int> A;

    for (int i = 0; i < n; i++)
    {
        string s;
        cin >> s;

        if (s == "PUSHBACK")
        {
            int idk;
            cin >> idk;
            A.push_back(idk);
        }
        else if (s == "PUSHFRONT")
        {
            int idk;
            cin >> idk;
            A.push_front(idk);
        }
        else if (s == "POPBACK")
        {
            if (!A.empty()) A.pop_back();
        }
        else if (s == "POPFRONT")
        {
            if (!A.empty()) A.pop_front();
        }
        else if (s == "PRINTBACK")
        {
            !A.empty() ? cout << A.back() : cout << "NONE";
            cout << endl;
        }
        else if (s == "PRINTFRONT")
        {
            !A.empty() ? cout << A.front() : cout << "NONE";
            cout << endl;
        }
    }
}

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    submit();

    return 0;
}

File: DSA_python/DSA08003 HÀNG ĐỢI HAI ĐẦU (DEQUEUE).py
def submit():
    n = int(input())
    A = list()

    for i in range(n):
        s = input().split()

        if s[0] == "PUSHBACK":
            A.append(s[1])
        elif s[0] == "POPBACK":
            if len(A) != 0: A.pop()
        elif s[0] == "PUSHFRONT":
            A.insert(0, s[1])
        elif s[0] == "POPFRONT":
            if len(A) != 0: A.remove(A[0])
        elif s[0] == "PRINTBACK":
            print(A[len(A)-1] if len(A) != 0 else "NONE")
        else:
            print(A[0] if len(A) != 0 else "NONE")


submit()
